### PR TITLE
chore: Bump `reselect` to `^5.1.1` for heterogeneously-typed selectors support

### PR DESCRIPTION
**Pull request by MajorLift**
#### Homogenous selector types:

```ts
const selectorOne = (state: State) => state.something;
const selectorTwo = (state: State) => state.other;
createSelector([selectorOne, selectorTwo], () => ...);
```

#### Heterogenous selector types: 

```ts
const selectorOne = (state: { something: string }) => state.something;
const selectorTwo = (state: { other: number }) => state.other;
createSelector([selectorOne, selectorTwo], () => ...);
```

Support for heterogeneous typing is essential for selectors to function properly, because selectors must be both mergeable and atomic. Without heterogeneous selectors, these becoming conflicting objectives.

- **Mergeable:** 
  - Without heterogeneous typing for selectors, the size of the state type for all selectors tend to inflate. This is because a selector's state must be a supertype of the intersection of the state types of all of its merged selectors, including selectors nested in the definitions of merged selectors. 
  - Eventually, many selectors end up being defined with a state type that is close to the entire Redux state. 
  - Paradoxically, selectors that only need access to very few properties end up needing to have the widest state type, because they tend to be merged into the most selectors across several nested levels. 
  
- **Atomic:**
  - When selectors are actually invoked, including in test files, it's not always practical to prepare and pass in a very large state object. 
  - It's both safer and more convenient to restrict the state being passed into the selector to the minimum size required for the selector to function. 
  - This requirement becomes incompatible with the mergeability requirement if all selectors must share a common state type. 
  
Enabling merged selectors to accept different, even disjoint state types resolves this issue.

> [!NOTE]
> See the [`MultichainState`](https://github.com/MetaMask/metamask-extension/blob/4f970df0acec3e3bc80da08373aa3b16f23aae41/ui/selectors/multichain.ts#L53) type for an example of a bloated selector state type which will become unnecessary with this update.

## **Description**

- `reselect@4.0.0` supports heterogeneous typing for selector inputs to `createSelector` and `createDeepEqualSelector`.
  - https://github.com/reduxjs/reselect/issues/351
    - https://github.com/reduxjs/reselect/pull/274
    - https://github.com/reduxjs/reselect/pull/315
- Upgrade to `^5.1.1` (latest) is necessary to fix type issues in `4.0.0`
  - https://app.circleci.com/jobs/github/MetaMask/metamask-extension/4320173

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/29094?quickstart=1)

## **Related issues**

- Blocks TypeScript conversion of selectors for https://github.com/MetaMask/MetaMask-planning/issues/2894.
  - https://github.com/MetaMask/metamask-extension/pull/29014

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by MajorLift** [comment]
@metamaskbot update-policies

File: shared/modules/selectors/util.js [hunk 1]
```diff
@@ -1,9 +1,9 @@
 import { TransactionStatus } from '@metamask/transaction-controller';
 import { isEqual } from 'lodash';
-import { createSelectorCreator, defaultMemoize } from 'reselect';
+import { createSelectorCreator, lruMemoize } from 'reselect';
 
 export const createDeepEqualSelector = createSelectorCreator(
-  defaultMemoize,
+  lruMemoize,
```
Comment (FrederikBolding) [inline shared/modules/selectors/util.js:6]: Just to clarify, this implementation is the same under the hood right?

**Review by FrederikBolding — APPROVED** [review]
Snaps changes LGTM

File: shared/modules/selectors/util.js (hunk 1 above)
Comment (MajorLift) [reply shared/modules/selectors/util.js:6]: Yes! It was just renamed in v5.

Comment (MajorLift) [reply shared/modules/selectors/util.js:6]: - Changelog: https://github.com/reduxjs/reselect/releases/tag/v5.0.1
- PR: https://github.com/reduxjs/reselect/pull/654

**Review by infiniteflower — APPROVED** [review]

**Pull request merged**